### PR TITLE
Vision transformer inception resnet

### PR DESCRIPTION
**Pull request by Jakoviz**
Vision transformer, inception and resnet merged with the train.py structure in main. Hyperparameter search with those three from the command line.

### CODE REVIEW

File: hyperparams.yaml [hunk 1]
```diff
@@ -1,10 +1,39 @@
...
   BATCH_SIZE_TRAIN: 64
   BATCH_SIZE_TEST: 64
   LR: 0.01
+  WEIGHT_DECAY: 0
+  DAMPENING: 0
+  MOMENTUM: 0
+  EPS: 0
+  BETAS: 0
+  LR_DECAY: 0
+  ALPHA: 0
```
Comment (perander) [inline hyperparams.yaml:13]:
```suggestion
  ALPHA: 0
  NUM_HEADS: 0
  DROPOUT: 0
```

File: hyperparams.yaml [hunk 2]
```diff
@@ -1,10 +1,39 @@
...
-  LR: 0.01
\ No newline at end of file
+  LR: 0.01
+  WEIGHT_DECAY: 0
+  DAMPENING: 0
+  MOMENTUM: 0
+  EPS: 0
+  BETAS: 0
+  LR_DECAY: 0
+  ALPHA: 0
```
Comment (perander) [inline hyperparams.yaml:26]:
```suggestion
  ALPHA: 0
  NUM_HEADS: 0
  DROPOUT: 0
```

File: hyperparams.yaml [hunk 3]
```diff
@@ -1,10 +1,39 @@
...
+  BATCH_SIZE_TRAIN: 64
+  BATCH_SIZE_TEST: 64
+  LR: 0.01
+  WEIGHT_DECAY: 0
+  DAMPENING: 0
+  MOMENTUM: 0
+  EPS: 0
+  BETAS: 0
+  LR_DECAY: 0
+  ALPHA: 0
```
Comment (perander) [inline hyperparams.yaml:39]:
```suggestion
  ALPHA: 0
  NUM_HEADS: 8
  DROPOUT: 0.2
```

File: train.py [hunk 4]
```diff
@@ -125,7 +127,30 @@ def train(model, dataset, data_csv, binary, params_file, augmentation, save, ver
         device = torch.device('cpu')
 
     model = get_model_class(model, num_of_classes=NUM_CLASSES).to(device)
```
Comment (perander) [inline train.py:129]:
```suggestion
        model = get_model_class(model, num_of_classes=NUM_CLASSES, num_heads=params[model]['NUM_HEADS'], dropout=params[model]['DROPOUT']
```

**Review by perander — COMMENTED** [review]
Added two hyperparameters specific to vision transformer.

File: hyperparameter_search.py [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+        avg_valid_losses.append(total_valid_loss / (batch_num + 1))
+        valid_accuracies.append(100. * total_correct / total)
+
+        if FLAG_EARLYSTOPPING:
+            # early_stopping needs the validation loss to check if it has decresed,
+            # and if it has, it will make a checkpoint of the current model
+            early_stopping(avg_valid_losses[-1], model)
+            # If the loss has not decreased for {patience} number of epochs, trigger early stop
+            if early_stopping.early_stop:
+                print("Early stop")
```
Comment (Redande) [inline hyperparameter_search.py:243]:
```suggestion
                logger.info("Early stop")
```

File: hyperparameter_search.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+# %%
+def compute_and_print_metrics(stage, NUM_CLASSES, epoch, total_correct, total, true_positive,
+        true_negative, false_positive, false_negative, target_all_batches, pred_all_batches, batch_num, no_of_batches, loss):
+    print(f"{stage}: Epoch {epoch} - Batch {batch_num + 1}/{no_of_batches}: Loss: {loss} | {stage} Acc: {100. * total_correct / total} ({total_correct}/{total})")
+    print(f"{stage} TP: {true_positive} TN: {true_negative} FP: {false_positive} FN: {false_negative}")
+    recall = true_positive / (true_positive + false_negative + 1e-10)
+    precision = true_positive / (true_positive + false_positive + 1e-10)
+    print(f"{stage} Recall: {recall}")
+    print(f"{stage} Precision: {precision}")
+    print(f"{stage} F1: {2 * precision * recall / (precision + recall + 1e-10)}")
```
Comment (Redande) [inline hyperparameter_search.py:41]:
```suggestion
    logger.info(f"{stage} Recall: {recall}")
    logger.info(f"{stage} Precision: {precision}")
    logger.info(f"{stage} F1: {2 * precision * recall / (precision + recall + 1e-10)}")
```

File: hyperparameter_search.py [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+load_dotenv()
+DATA_FOLDER_PATH = os.getenv("DATA_FOLDER_PATH")
+PLANT_SPLIT_MASTER_PATH = os.path.join(DATA_FOLDER_PATH, "plant_data_split_master.csv")
+warnings.filterwarnings("ignore")
+
+# %%
+def compute_and_print_metrics(stage, NUM_CLASSES, epoch, total_correct, total, true_positive,
+        true_negative, false_positive, false_negative, target_all_batches, pred_all_batches, batch_num, no_of_batches, loss):
+    print(f"{stage}: Epoch {epoch} - Batch {batch_num + 1}/{no_of_batches}: Loss: {loss} | {stage} Acc: {100. * total_correct / total} ({total_correct}/{total})")
+    print(f"{stage} TP: {true_positive} TN: {true_negative} FP: {false_positive} FN: {false_negative}")
```
Comment (Redande) [inline hyperparameter_search.py:36]:
```suggestion
    logger.info(f"{stage}: Epoch {epoch} - Batch {batch_num + 1}/{no_of_batches}: Loss: {loss} | {stage} Acc: {100. * total_correct / total} ({total_correct}/{total})")
    logger.info(f"{stage} TP: {true_positive} TN: {true_negative} FP: {false_positive} FN: {false_negative}")
```

File: hyperparameter_search.py [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+    print(f"{stage} Precision: {precision}")
+    print(f"{stage} F1: {2 * precision * recall / (precision + recall + 1e-10)}")
+
+    f1m = f1_score(target_all_batches.detach().cpu(), pred_all_batches.detach().cpu(), average = 'macro', zero_division=1)
+    f1w = f1_score(target_all_batches.detach().cpu(), pred_all_batches.detach().cpu(), average = 'weighted', zero_division=1)
+    f1mi = f1_score(target_all_batches.detach().cpu(), pred_all_batches.detach().cpu(), average = 'micro', zero_division=1)
+
+    print(f"{stage} unweighted macro F1 as per sklearn: {f1m}")
+    print(f"{stage} weighted macro F1 as per sklearn: {f1w}")
+    print(f"{stage} 'global' micro F1 as per sklearn: {f1mi}")
```
Comment (Redande) [inline hyperparameter_search.py:49]:
```suggestion
    logger.info(f"{stage} unweighted macro F1 as per sklearn: {f1m}")
    logger.info(f"{stage} weighted macro F1 as per sklearn: {f1w}")
    logger.info(f"{stage} 'global' micro F1 as per sklearn: {f1mi}")
```

File: hyperparameter_search.py [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+
+    if (NUM_CLASSES == 2):
+        f1b = f1_score(target_all_batches.detach().cpu(), pred_all_batches.detach().cpu(), average = 'binary', zero_division=1)
+        print(f"{stage} binary F1 as per sklearn: {f1b}")
```
Comment (Redande) [inline hyperparameter_search.py:53]:
```suggestion
        logger.info(f"{stage} binary F1 as per sklearn: {f1b}")
```

File: models/model_parts.py [hunk 10]
```diff
@@ -60,4 +60,37 @@ def forward(self, x):
...
+            nn.Linear(hidden_dim, embed_dim),
+            nn.Dropout(dropout)
+        )
+
+
+    def forward(self, x):
+        inp_x = self.layer_norm_1(x)
+        x = x + self.attn(inp_x, inp_x, inp_x)[0]
+        x = x + self.linear(self.layer_norm_2(x))
+        return x
```
Comment (Redande) [inline models/model_parts.py:96]:
```suggestion
        return x

```

File: models/vision_transformer.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    return VisionTransformer(
+        num_classes=num_classes, 
+        embed_dim=256, 
+        hidden_dim=512,
+        num_channels=3, 
+        num_heads=num_heads,
+        num_layers=6, 
+        patch_size=32, 
+        num_patches=64,
+        dropout=dropout)
```
Comment (Redande) [inline models/vision_transformer.py:76]:
```suggestion
        dropout=dropout)

```

File: utils/image_utils.py [hunk 12]
```diff
@@ -137,3 +139,36 @@ def convert_color_to_another(image, color1=[255, 255, 255], color2=[0, 0, 0]):
...
+
+  fig, ax = plt.subplots(images.shape[0], 1, figsize=(14,3))
+  fig.suptitle("Images as input sequences of patches")
+  for i in range(images.shape[0]):
+      img_grid = torchvision.utils.make_grid(img_patches[i], nrow=64, normalize=True, pad_value=0.9)
+      img_grid = img_grid.permute(1, 2, 0)
+      ax[i].imshow(img_grid)
+      ax[i].axis('off')
+  plt.show()
+  plt.close()
```
Comment (Redande) [inline utils/image_utils.py:174]:
```suggestion
    images = torch.stack([dataset[idx]['image'] for idx in range(num_images)], dim=0)

    img_patches = img_to_patch(images, patch_size=32, flatten_channels=False)

    fig, ax = plt.subplots(images.shape[0], 1, figsize=(14,3))
    fig.suptitle("Images as input sequences of patches")
    for i in range(images.shape[0]):
        img_grid = torchvision.utils.make_grid(img_patches[i], nrow=64, normalize=True, pad_value=0.9)
        img_grid = img_grid.permute(1, 2, 0)
        ax[i].imshow(img_grid)
        ax[i].axis('off')
    plt.show()
    plt.close()

```

Comment (Redande) [reply utils/image_utils.py:174]: Fixes indentation and adds a newline to the end of file

**Review by Redande — APPROVED** [review]
Looks good! I added some minor modification suggestions to conform to the styling in other parts of our code

**Review by jarsba — COMMENTED** [review]
Looks good!

**Review by jarsba — APPROVED** [review]

**Pull request merged**